flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/ClusterEntrypoint.java

```diff
@@ -558,9 +587,29 @@ private Configuration generateClusterConfiguration(Configuration configuration)
      * @throws IOException if the temporary directories could not be cleaned up
      */
     protected void cleanupDirectories() throws IOException {
+        IOException ioException = null;
+
         final String webTmpDir = configuration.getString(WebOptions.TMP_DIR);
```

Comment:
I think the `webTmpDir` is currently being cleaned up. However, it does not live in the working dir. As a follow-up we could move it to the working directory.

Reply:
The issue is that it is _not_ cleaned up on PROCESS_FAILURE, and a restarted JM has no knowledge of the previous `webTmpDir` (so we're just leaking it)..
Even if the webTmpDir were to be placed by default in the working directory, it should still always be cleaned up on the off-chance that a user has explicitly configured WebOptions.TMP_DIR.

Reply:
I think you are right. I will change it so that we pass in the shutdown behaviour to the `cleanupDirectories` method and then only delete the working directory if we are in `GRACEFUL_SHUTDOWN`.